Add access to GameProfile cache, a GameProfile resolver and Server.getUser(UUID)

boformer opened the pull request:
This PR adds a service interface that can be used to lookup player UUIDs by name and player names by UUIDS. The service contacts the Mojang auth servers if no cached profile was found in the local cache.
It also adds a ``UserStorage`` interface to get, create or delete a ``User`` object.

This resolves https://github.com/SpongePowered/SpongeAPI/issues/528

Access to cached offline user data is important to:

* complete or validate names of offline players
* lookup the name of an offline player by an UUID stored by a plugin
* ban offline players

Things to discuss:

**Access to the [Name Changes API](http://wiki.vg/Mojang_API#UUID_-.3E_Name_history)?** 
Mojang provides an API which lists up all names and change dates of a player.

Minecrell commented:
Cache isn't really a good name for getting `User` instances. The difference between `GameProfile` and `User` is not only persistence or being cached, they represent two completely different use cases:

- `GameProfile`: The identification of a player (UUID + name), as well as optionally a few properties from Mojang's server (like skin and cape). This is generally not persistent, but Minecraft is maintaining a cache for the game profiles for faster profile lookup (`usercache.json`). This includes not only players that have played on the server, but also name->UUID lookups caused through commands.

- `User`: Persistent data of a player, this contains his location in worlds, as well as inventory contents and all the other persistent data of a `Player` - even though not everything is accessible through SpongeAPI. For that to exist, the player has to join the server once. This gets saved together with the world data, in NBT.

A cache is typically used for performance, where entries (can) expire and where it doesn't matter much if certain entries get lost, because they can be queried again. While you could see `User` as cached data being loaded from the disk, a "cache" would be more appropriate for `GameProfile`s, which will be eventually queried from Mojang servers.

Loading `User` data should never trigger any queries to Mojang's server, but rather rely only on the saved persistent user data. If you would want to get a player's data with his newer name (e.g. because he hasn't joined the server since he changed yet) you would need to query name->UUID using a `GameProfile` first and then load the `User` using the returned UUID.

boformer commented:
@Minecrell What I primarily want is to access the local username<-->uuid storage (only for players who played on the server before).

What do you think of this proposal?

* UserStorage (maybe part of ``Server``)
  * get ``User`` by UUID
* GameProfileCache
  * get stored ``GameProfile`` by UUID
. * get stored ``GameProfile`` by name
. * get list of all profiles
. * list of matching profiles starting with string
* GameProfileResolver
  * get ``GameProfile`` from auth server by UUID
. * get ``GameProfile`` from auth server by name

Minecrell commented:
@boformer Basically yes, although I think `getUser` should be actually at the same place as `getPlayer` so it would be probably a good idea to move it.

boformer commented:
I added a new commit which adds the profile cache and resolver and a few new methods in the server interface.

simon816 commented on src/main/java/org/spongepowered/api/service/userdata/UserDataCache.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return A list of cached users
+     */
+    List<User> getAll();
```
I would imagine this method could be expensive, especially for a lot of previous player data. Perhaps it could be `List<UUID>` instead?

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileCache.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return A collection of {@link GameProfile}s
+     */
+    ListenableFuture<Collection<GameProfile>> getAll();
```
Why is this a future? The cached game profiles should be available without additional query time

boformer replied:
It's based on this suggestion: https://github.com/SpongePowered/SpongeAPI/issues/528#issuecomment-94164268

In theory, the cache could also be a database, so both service interfaces extend ``GameProfileSource``, which also utilizes ``Future``s.

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileSource.java:
```diff
@@ -0,0 +1,71 @@
+     * @param uniqueId The unique ID
+     * @return The result of the request
+     */
+    ListenableFuture<Optional<GameProfile>> get(UUID uniqueId);
```
I think we can generally say, fetch the game profile or fail (with an exception), so the Optional is not needed

boformer commented:
Some more things to discuss:

**Access to the [Name Changes API](http://wiki.vg/Mojang_API#UUID_-.3E_Name_history)?** 
Mojang provides an API which lists up all names and change dates of a player.

**Server.deleteUser(..) method?** 

**A chained cache --> resolver interface?** 
First check the cache, contact Mojang servers if no data is cached or is expired?

ST-DDT commented:
> get list of all profiles (boolean playedBefore to exclude lookups)

What kind of lookup? I guess you are not refering to get all available GameProfiles from Mojang.

boformer commented:
No, from the local profile cache.
The cache may also store profiles of players who never played on the server (a plugin just queried the mojang server, and the result is stored in cache).

For the command autocomplete feature, these profiles should not be included.

The completer could also check if a ``User`` object exists for a profile.

To workflow to get a user by name is just becoming very complicated:
* check if profile is in cache
  * no? Query mojang server 
    * no? --> Profile does not exist
* get User object for profile
  * absent? --> player never played on server
* check if user hasPlayedBefore()
...

boformer commented:
I thought about this again:

* It would make sense to merge the resolver and the cache. Caching should be handled internally (expiration date)
* We need a function that returns a collection of all Users with a name that starts with a string.

boformer commented:
* The function should not cause a http request to a Mojang server, but use the last known username.

:arrow_right:  We have to store the last known username of a ``User``.

MattBDev commented:
Any updates on the status of this PR?

boformer commented:
@MattBDev I think it is fleshed out now. It basically adds 2 new service interfaces:

``UserStorage`` - To get a ``User`` by UUID or last known name
``GameProfileResolver`` - Search the profile cache or contact Mojang servers to get the current name by UUID (or vice versa)

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,83 @@
+ * an {@link IOException} if a network error occured.</p>
+ */
+public interface GameProfileResolver
+{
```
Code formatting is wrong here

simon816 commented on src/main/java/org/spongepowered/api/Server.java:
```diff
@@ -25,8 +25,10 @@
+
 import org.spongepowered.api.entity.Entity;
 import org.spongepowered.api.entity.player.Player;
+import org.spongepowered.api.entity.player.User;
```
Unused import

me4502 replied:
All of the modifications in this file are unrelated to this PR.

MattBDev replied:
Changes on this file should be removed before pulling.

simon816 commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,83 @@
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * A service which contacts the authentification servers to get a
```
Spelling. "authentification" => "authentication"

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,104 @@
+     * Deletes the data associated with a {@link User}.
+     *
+     * @param profile The profile
+     * @return The user object
```
this is a `void` method, no `@return`. I suspect this was copy + paste from above method.

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,104 @@
+     * Returns a collection of matching {@code User}s whose last known user
+     * names start with the given string (case-insensitive).
+     *
+     * @param name
```
`@param lastKnownName`, also need description of parameter

simon816 commented:
Would this PR be an appropriate place to move the method [`createGameProfile`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameRegistry.java#L366) out of `GameRegistry` as I don't think it belongs there?

MattBDev commented:
@simon816 I think that method may have a use that these new classes won't work (I can't think of any myself though) for although I could be wrong.

boformer commented:
@simon816 The method in GameRegistry has it's own use cases:

* Import ``User`` data from another server (e.g. a Bukkit server)
* Add fake profiles to a ``StatusPingEvent``

MattBDev commented:
I think if all the changes on the Server.java file are removed, this can be pulled.

simon816 commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,83 @@
+ * <p>The service may cache the data of a request for faster lookups. Note that
+ * the cached data may not always be up to date.</p>
+ *
+ * <p>The returned {@code ListenableFuture} throws an {@link ExecutionException}
```
`{@link ListenableFuture}`

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,103 @@
+    Collection<User> getAll();
+
+    /**
+     * Returns a collection of matching {@code User}s whose last known user
```
`{@link User}`

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Gets the collection of all stored {@link User}s.
+     *
+     * @return A {@link Collection} of users.
```
Remove the period

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Gets the data of a {@link User} by their profile.
+     *
+     * @param uniqueId The UUID of the user.
```
Remove the period

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,103 @@
+     * <p>To get the current name of a player, use the
+     * {@link GameProfileResolver} service.</p>
+     *
+     * @param uniqueId The UUID of the user.
```
Remove the period

simon816 commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Gets the data of a {@link User} by their unique id.
+     *
+     * @param uniqueId The UUID of the user.
```
Remove the period

simon816 commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,83 @@
+     * Gets a collection of {@link GameProfile}s by their user names
+     * (case-insensitive).
+     *
+     * @param names The user names
```
Need `@param useCache`

simon816 commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * Looks up a {@link GameProfile} by its user name (case-insensitive).
+     *
+     * @param name The username
```
Need `@param useCache`

MattBDev commented:
@boformer are you able to update the PR with the changes @simon816 suggested

boformer commented:
@MattBDev Done.

simon816 commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,85 @@
+ */
+package org.spongepowered.api.service.profile;
+
+import java.io.IOException;
```
Wrong import order, `java` imports should be below others. (See [guidelines](https://docs.spongepowered.org/en/contributing/guidelines.html))

boformer commented:
@simon816 I finally found out how to control this with eclipse

MattBDev commented:
:exclamation: *waits impatiently for sponge team to review this*

RobertHerhold commented:
@MattBDev 
Just chill out for a minute. These guys have full-time jobs, and they are *very* busy at times. They will review this when they get to it.

Zidane commented:
@SpongePowered/developers 

Final call before this is merged. I need some feedback.

me4502 commented:
looks good to me

Minecrell commented:
How would you resolve a `GameProfile` **only** from the cache (if available), without querying Mojang's servers? Sometimes it might be not that important so you just want it if available.

Minecrell commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @param profile The profile
+     */
+    void delete(GameProfile profile);
```
What would happen here if the user is currently online?

MattBDev replied:
I think it would make sense to delete inventory and other data if they are online

simon816 replied:
It would cause sync issues with the client unless everything is updated on the client. Location, inventory, hunger, health etc. Maybe this should fail (and return a boolean false) if the user is online.

boformer replied:
Or throw an exception?

MattBDev replied:
An exception would work.

ryantheleach replied:
The API has largely avoided checked exceptions where possible, why add one here, now?

Minecrell commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,102 @@
+     * @param lastKnownName The user name
+     * @return The result of the request
+     */
+    Collection<User> match(String lastKnownName);
```
Couldn't we provide the same also for `GameProfile`s? There will be most likely more `GameProfile`s cached than `User`s, so you could provide more profiles than users (e.g. when used for tab completion).

boformer replied:
The problem is that cached gameprofiles expire (at least in vanilla). Also, not all cached gameprofiles are players who played on the server.

boformer replied:
@Minecrell Do you think this method should also return GameProfiles for performance reasons? At least that would be sufficient for tab completion.

Minecrell replied:
If you change the other method this should also return `GameProfile`, yes.

Minecrell commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return A {@link Collection} of users
+     */
+    Collection<User> getAll();
```
Considering that this can be quite inefficient for bigger uses (which may have 5,000+ players persisted) this should have _at least_ some note that this can be quite inefficient depending on the player count on the server.

simon816 replied:
I think it should return a collection of UUIDs as it is much cheaper to load many UUID objects.

me4502 replied:
I feel GameProfile would be more useful.

boformer replied:
Ok, I will change it to GameProfile.

ryantheleach replied:
If it's that likely to be inefficient, maybe make some sort of a lazy collection, or provide search methods or some way to page it?

Minecrell commented:
Can you remove [`GameProfile.getUser()`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameProfile.java#L43-L48)? With the additions in here these methods should be preferred over the current one.

boformer commented:
@minecrell yes, I will. Also thought about that today.

boformer commented:
I have removed the getUser method from GameProfile and changed the return types of the resource-intensive methods to Collection<GameProfile>. I also changed the return type of the delete method to boolean to indicate the success of the deletion.

I think this is ready.

Minecrell commented:
I'll repeat this question from above:
> How would you resolve a `GameProfile` **only** from the cache (if available), without querying Mojang's servers? Sometimes it might be not that important so you just want it if available.

boformer commented:
@Minecrell If you want to lookup a player who played on the server before, you can use getUser(UUID/Name).getProfile(), though that's not very elegant.

Do you think it is necessary to have direct access to the profile cache (which also contains players who did not play on the server)?

I could add a third service for that: ``GameProfileCache``, similar to ``GameProfileResolver`` (maybe without Futures).

Or replace the ``boolean useCache`` parameter with a ``ResolveMethod`` enum parameter: CACHE, REMOTE, CACHE_REMOTE.

What do you prefer?

Minecrell commented:
I would basically like a way to access the cache _without_ any delay (-> without a future). Sometimes you don't want (or can't) halt the execution of your code to wait for the exectution of a future. IMO there should also be a way to access only the cached profiles with direct access.

I know that you made it that way because of https://github.com/SpongePowered/SpongeAPI/pull/580#discussion_r28693330, however there will be most likely a local memory cache anyway simply because that's much faster than querying a remote server everytime. And if not, this cache could simply never return anything.

Minecrell commented:
@boformer Actually never mind, I realized there is really not a big use case for it.

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,84 @@
+     * @param useCache true to perform a cache lookup first
+     * @return The result of the request
+     */
+    ListenableFuture<GameProfile> get(UUID uniqueId, boolean useCache);
```
What do you think about separating this in like `get()` and `query()` instead of the parameter,  it makes reading the code more clear: `resolver.query(uuid)` vs `resolver.get(uuid, false)` (you always have to look what exactly was this boolean again)

boformer replied:
Or just ``get(UUID)`` (Shortcut for ``get(UUID, true)``)and ``get(UUID, boolean)``

ST-DDT replied:
I prefer get and query.
Because the difference between them is very clear due to the method naming.

boformer replied:
I'm not sure about this. Mojang recommends to cache profiles.
I don't want to provide a shortcut for "contact servers directly".

That's why i prefer 2 methods with the same name. Only developers who really want to force a server lookup will use the boolean function.

progwml6 replied:
The cache bypass is not the best thing to let people near

MattBDev replied:
Agreed. It's best to have it default to true because performing too many direct requests to Mojang servers can cause the the server to be blacklisted from accessing the API.
Most developers will most likely prefer to use `get(UUID uniqueId)` which will be a lot safer and avoid this risk.

Minecrell commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,115 @@
+     * @param profile The profile
+     * @return true if the deletion was successful
+     */
+    boolean delete(GameProfile profile);
```
Couldn't this take a user? This would ensure the given parameter is really a user that exists.

boformer replied:
A ``User`` object could also be deleted (if you call the method twice).

I will just add both methods...

Minecrell commented on src/main/java/org/spongepowered/api/service/user/UserStorage.java:
```diff
@@ -0,0 +1,115 @@
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAll();
```
While these both methods return a `GameProfile` already, I think they would also additionally make sense with the data from the cache. For example for a command adding player heads to the inventory you don't really care if the user has played on the server already, you want too provide as many options in the tab complete as possible.

boformer replied:
For that use case, we should add similar methods to the profile resolver service:

``getAllCachedProfiles()`` and ``matchCachedProfile(String)``

Minecrell commented:
You have a checkstyle warnings in here https://travis-ci.org/SpongePowered/SpongeAPI/jobs/62137985#L257

MattBDev commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,135 @@
+     *
+     * <p>This collection may also contain profiles of players who never played
+     * on the server!</p>
+     *
```
If this only fetches from the local cache then how can it contain profiles of players who have never played?

boformer replied:
Whenever you do a lookup with get(UUID/name), the profile is saved in the local cache.

You can also lookup profiles of players who never played on the server.

MattBDev replied:
Thanks for this clarification.

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAllCachedProfiles();
```
Why not just `getAll`?

Minecrell commented:
https://travis-ci.org/SpongePowered/SpongeAPI/jobs/62295585#L264

boformer commented:
I squashed the commits.

MattBDev commented:
:heart_eyes: Looks awesome

boformer commented:
I fixed the travis errors and renamed the method to ``getAllCached()``.

Minecrell commented on src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAllCached();
```
I think this should be either `getCachedProfiles()` or `getAll()` but `getAllCached()` sounds a bit weird

boformer commented:
I renamed the method to ``getAllCachedProfiles`` to make clear what it does. ``getAll`` is misleading.

Minecrell commented:
@boformer `getAllCachedProfiles` is essentially the same as `getCachedProfiles`, the all is kinda redundant in the name.

boformer commented:
You are right.

gabizou commented on src/main/java/org/spongepowered/api/service/profile/ProfileNotFoundException.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Thrown by a profile lookup service if a profile does not exist.
+ */
+public class ProfileNotFoundException extends Exception {
```
Couldn't this extend some other exception than just `Exception`?

gabizou replied:
Never mind. It's fine to just extend `Exception` in this case.

simon816 commented:
Why is this a service? I don't think it should be possible for plugins to replace the implementation of the  `GameProfileResolver` and `UserStorage`.

MattBDev commented:
I didn't realize that this was being added as a service. Extra plugins just to get a User shouldn't be needed. This should be only implemented by sponge

boformer commented:
@simon816 First of all, almost any system is a service, even the command and event system.

Maybe it would even make sense to have a custom GameProfileResolver with a custom caching solution etc.

modwizcode commented:
That may just represent a more fundamental problem with the way services
are being used.

On Thu, May 14, 2015 at 4:23 PM, boformer <notifications@github.com> wrote:

> @simon816 <https://github.com/simon816> First of all, almost any system
> is a service, even the command and event system.
>
> Maybe it would even make sense to have a custom GameProfileResolver with a
> custom caching solution etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/580#issuecomment-102175101>
> .
>



-- 

Starbuck Johnson, Java Developer
www.modwiz.com | starbuck@modwiz.com

boformer commented:
@modwizcode Services are a part of the Sponge philosophy. There are certain core services that can't be overridden. I don't see a problem with that.

modwizcode commented:
Alright I guess I misunderstood the goal of services.

On Thu, May 14, 2015 at 4:41 PM, boformer <notifications@github.com> wrote:

> @modwizcode <https://github.com/modwizcode> Services are a part of the
> Sponge philosophy. There are certain core services that can't be
> overridden. I don't see a problem with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/580#issuecomment-102178139>
> .
>



-- 

Starbuck Johnson, Java Developer
www.modwiz.com | starbuck@modwiz.com

ryantheleach commented:
If it were possible for UserStorage to be overridden it would be awesome.

Store all the users out to a database instead, and you already have cross server inventory compatibility.

simon816 commented:
If plugins were able to implement this, they would need to provide an implementation of `User`, which of course would be a problem as the implementation will be expecting it to be an instance of their own user (say `SpongeUser`), ultimately leading to cast exceptions.

MattBDev commented:
I think if anyone was to implement this in their plugin, they would do so only if they knew what they were doing. I retract my previous statement of not adding this as a service.
